django/applications/catmaid/migrations/0006_add_history_table_support.py

```diff
@@ -0,0 +1,893 @@
+                    txid_current(),
+                    string_agg('ht.' || quote_ident(c.column_name) || '<>' ||
+                        'lt.' || quote_ident(c.column_name), ' OR '),
+                    pkey_column
+                )
+                FROM information_schema.columns c, pg_class pc
+                WHERE pc.oid = live_table_name
+                AND c.table_name = pc.relname
+            );
+
+            GET DIAGNOSTICS num_updated_rows = ROW_COUNT;
+
+            IF num_updated_rows > 0 THEN
+                RAISE NOTICE '% existing history entries required an update',
+                    num_updated_rows;
+            ELSE
+                RAISE NOTICE 'No existing history entries required an update';
+            END IF;
+        END IF;
+
+        -- Insert all live rows that don't have an existing history entry yet.
+        EXECUTE (
+            SELECT format (
+                'INSERT INTO %1$s (%5$s,sys_period,exec_transaction_id) '
+                'SELECT %4$s, tstzrange(%3$s, null), %6$s '
+                'FROM %2$s lt '
+                'LEFT JOIN %1$s ht '
+                'ON lt.%7$s = ht.%7$s '
+                'WHERE ht.%7$s IS NULL',
+                history_table_name,
+                live_table_name,
+                time_source,
+                string_agg('lt.' || quote_ident(c.column_name), ','),
+                string_agg(quote_ident(c.column_name), ','),
+                txid_current(),
+                pkey_column
+            )
+            FROM information_schema.columns c, pg_class pc
+            WHERE pc.oid = live_table_name
+            AND c.table_name = pc.relname
+        );
+
+        GET DIAGNOSTICS num_new_rows = ROW_COUNT;
+
+        IF num_new_rows > 0 THEN
+            RAISE NOTICE '% new live entries where added to the history',
+                num_new_rows;
+        ELSE
+            RAISE NOTICE 'No new live entries where added to the history';
+        END IF;
+
+        end_time = clock_timestamp();
+        delta = 1000 * (extract(epoch from end_time) - extract(epoch from start_time));
+        RAISE NOTICE 'Execution time: %ms', delta;
+    END;
+    $$;
+
+
+    -- Remove an existing history table for the passed in table
+    CREATE OR REPLACE FUNCTION drop_history_table(live_table_name regclass)
+    RETURNS void
+    LANGUAGE plpgsql AS
+    $$
+    DECLARE
+
+        -- This will contain the name of the newly created history table. No
+        -- regclass is used, because the implicit table existence check on variable
+        -- assignment can fail if the table has already been removed by an
+        -- cascaded table drop.
+        history_table_name text;
+
+    BEGIN
+
+        -- History tables will be named like the live table plus a '_history' suffix
+        history_table_name = history_table_name(live_table_name);
+
+        -- Cascading deleting is used to delete parent tables and child tables in one go
```

Comment:
You are of course right, it does *not* cascade to parent tables. What I meant with my comment was that not only the table itself but also its children (to which the table in focus is a parent) are dropped. I fixed the comment. History tables for parent tables have to be dropped separately.